File: android/src/main/kotlin/com/poilink/sdk/flutter/PoilinkSdkPlugin.kt
package com.poilink.sdk.flutter

import android.app.Activity
import android.os.Handler
import android.os.Looper
import com.poilink.sdk.CycleType
import com.poilink.sdk.MissionData
import com.poilink.sdk.MissionListFilter
import com.poilink.sdk.RewardType
import com.poilink.sdk.PendingItemGrant
import com.poilink.sdk.PoilinkSDK
import com.poilink.sdk.ProgressMissionMode
import com.poilink.sdk.SyncItemGrantsResult
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding

class PoilinkSdkPlugin : FlutterPlugin, ActivityAware, PoilinkHostApi {
  private var eventApi: PoilinkEventApi? = null
  private var activity: Activity? = null
  private val mainHandler = Handler(Looper.getMainLooper())

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    PoilinkHostApi.setUp(binding.binaryMessenger, this)
    eventApi = PoilinkEventApi(binding.binaryMessenger)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    PoilinkHostApi.setUp(binding.binaryMessenger, null)
    eventApi = null
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivity() {
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun setConfig(clientId: String, clientSecret: String) {
    PoilinkSDK.setConfig(clientId, clientSecret)
  }

  override fun initialize(callback: (Result<Unit>) -> Unit) {
    val act = activity
    if (act == null) {
      callback(Result.failure(FlutterError("1002", "Activity is not available", null)))
      return
    }
    PoilinkSDK.initialize(act,
      { callback(Result.success(Unit)) },
      { code, msg -> callback(Result.failure(FlutterError(code.toString(), msg, null))) }
    )
  }

  override fun authenticate(appUserId: String, callback: (Result<Unit>) -> Unit) {
    PoilinkSDK.authenticate(appUserId,
      { callback(Result.success(Unit)) },
      { code, msg -> callback(Result.failure(FlutterError(code.toString(), msg, null))) }
    )
  }

  override fun unauthenticate() {
    PoilinkSDK.unauthenticate()
  }

  override fun getRefreshToken(callback: (Result<String>) -> Unit) {
    PoilinkSDK.getRefreshToken(
      { token -> callback(Result.success(token)) },
      { code, msg -> callback(Result.failure(FlutterError(code.toString(), msg, null))) }
    )
  }

  override fun setRefreshToken(appUserId: String, refreshToken: String, callback: (Result<Unit>) -> Unit) {
    PoilinkSDK.setRefreshToken(appUserId, refreshToken,
      { callback(Result.success(Unit)) },
      { code, msg -> callback(Result.failure(FlutterError(code.toString(), msg, null))) }
    )
  }

  override fun progressMission(handle: Long, missionCode: String, amount: Long, mode: Long) {
    PoilinkSDK.progressMission(missionCode, amount.toInt(), ProgressMissionMode.fromValue(mode.toInt())) { code, msg ->
      runOnMain { eventApi?.onProgressMissionError(handle, code.toLong(), msg) { } }
    }
  }

  override fun progressMissionImmediate(
    missionCode: String,
    amount: Long,
    mode: Long,
    idempotencyKey: String?,
    callback: (Result<List<MissionDataMsg>>) -> Unit
  ) {
    PoilinkSDK.progressMissionImmediate(missionCode, amount.toInt(), ProgressMissionMode.fromValue(mode.toInt()),
      { list -> callback(Result.success(list.map(::toMissionMsg))) },
      { code, msg -> callback(Result.failure(FlutterError(code.toString(), msg, null))) },
      idempotencyKey
    )
  }

  override fun getMissionList(filter: MissionListFilterMsg?, callback: (Result<List<MissionDataMsg>>) -> Unit) {
    val nativeFilter = filter?.let {
      val f = MissionListFilter()
      f.cycleType = CycleType.fromValue((it.cycleType ?: 0).toInt())
      f.rewardType = RewardType.fromValue((it.rewardType ?: 0).toInt())
      val code = it.progressCode
      if (!code.isNullOrEmpty()) f.progressCode = code
      f
    }
    val list = if (nativeFilter != null) PoilinkSDK.getMissionList(nativeFilter) else PoilinkSDK.getMissionList()
    callback(Result.success(list.map(::toMissionMsg)))
  }

  override fun showWebPortal(options: WebPortalOptionsMsg) {
    PoilinkSDK.showWebPortal(buildWebPortalOptions(options))
  }

  override fun preloadWebPortal(options: WebPortalOptionsMsg) {
    PoilinkSDK.preloadWebPortal(buildWebPortalOptions(options))
  }

  override fun closeWebPortal() {
    PoilinkSDK.closeWebPortal()
  }

  override fun syncItemGrants(handle: Long) {
    PoilinkSDK.syncItemGrants(
      { grants -> runOnMain { eventApi?.onGrantsPage(handle, grants.map(::toGrantMsg)) { } } },
      { result -> runOnMain { eventApi?.onSyncComplete(handle, toResultMsg(result)) { } } },
      { code, msg -> runOnMain { eventApi?.onSyncError(handle, code.toLong(), msg) { } } }
    )
  }

  private fun buildWebPortalOptions(msg: WebPortalOptionsMsg): PoilinkSDK.WebPortalOptions {
    val handle = msg.handle ?: 0L
    val opts = PoilinkSDK.WebPortalOptions()
    opts.volume = msg.volume?.toFloat()
    opts.showMode = PoilinkSDK.ShowMode.fromValue((msg.showMode ?: 0).toInt())
    val x = msg.embeddedX
    val y = msg.embeddedY
    val w = msg.embeddedWidth
    val h = msg.embeddedHeight
    if (x != null && y != null && w != null && h != null) {
      opts.embeddedFrame = PoilinkSDK.EmbeddedFrame(x.toInt(), y.toInt(), w.toInt(), h.toInt())
    }
    opts.onShown = PoilinkSDK.SuccessCallback { runOnMain { eventApi?.onWebPortalShown(handle) { } } }
    opts.onClose = PoilinkSDK.SuccessCallback { runOnMain { eventApi?.onWebPortalClose(handle) { } } }
    opts.onError = PoilinkSDK.ErrorCallback { code, msg2 ->
      runOnMain { eventApi?.onWebPortalError(handle, code.toLong(), msg2) { } }
    }
    opts.onMissionChallenge = PoilinkSDK.MissionChallengeCallback { missionId ->
      runOnMain { eventApi?.onMissionChallenge(handle, missionId) { } }
    }
    opts.onRewardReceive = PoilinkSDK.RewardReceiveCallback { grantId, itemCode, quantity ->
      runOnMain { eventApi?.onRewardReceive(handle, grantId, itemCode, quantity.toLong()) { } }
    }
    return opts
  }

  // Android の SDK callback は呼び出しスレッドが一定でないため main に集約 (iOS は main 保証で不要)
  private fun runOnMain(block: () -> Unit) {
    mainHandler.post(block)
  }
}

private fun toMissionMsg(m: MissionData): MissionDataMsg = MissionDataMsg(
  inAppMissionId = m.inAppMissionId,
  progressCode = m.progressCode,
  title = m.title,
  details = m.details,
  point = m.point,
  targetProgress = m.targetProgress.toLong(),
  currentProgress = m.currentProgress.toLong(),
  hasAchievement = m.hasAchievement,
  isClaimed = m.isClaimed,
  displayOrder = m.displayOrder.toLong(),
  cycleType = m.cycleType.value.toLong(),
  rewardType = m.rewardType.value.toLong(),
  rewardItemCode = m.rewardItemCode,
  rewardItemQuantity = m.rewardItemQuantity.toLong()
)

private fun toGrantMsg(m: PendingItemGrant): PendingItemGrantMsg = PendingItemGrantMsg(
  grantId = m.grantId,
  itemCode = m.itemCode,
  quantity = m.quantity.toLong(),
  grantedAtUnixMs = m.grantedAtUnixMs
)

private fun toResultMsg(m: SyncItemGrantsResult): SyncItemGrantsResultMsg = SyncItemGrantsResultMsg(
  totalSynced = m.totalSynced.toLong(),
  totalMarked = m.totalMarked.toLong(),
  pageCount = m.pageCount.toLong()
)

File: ios/Classes/PoilinkSdkPlugin.swift
import Flutter
import UIKit
import poilink_sdk

public class PoilinkSdkPlugin: NSObject, FlutterPlugin {
  private static var instances: [PoilinkHostApiImpl] = []

  public static func register(with registrar: FlutterPluginRegistrar) {
    let messenger = registrar.messenger()
    let impl = PoilinkHostApiImpl()
    impl.eventApi = PoilinkEventApi(binaryMessenger: messenger)
    PoilinkHostApiSetup.setUp(binaryMessenger: messenger, api: impl)
    instances.append(impl)
  }
}

class PoilinkHostApiImpl: PoilinkHostApi {
  var eventApi: PoilinkEventApi?

  func setConfig(clientId: String, clientSecret: String) throws {
    PoilinkSDK.setConfig(clientId: clientId, clientSecret: clientSecret)
  }

  func initialize(completion: @escaping (Result<Void, Error>) -> Void) {
    PoilinkSDK.initialize(
      onSuccess: { completion(.success(())) },
      onError: { code, msg in completion(.failure(flutterError(code, msg))) }
    )
  }

  func authenticate(appUserId: String, completion: @escaping (Result<Void, Error>) -> Void) {
    PoilinkSDK.authenticate(
      appUserId: appUserId,
      onSuccess: { completion(.success(())) },
      onError: { code, msg in completion(.failure(flutterError(code, msg))) }
    )
  }

  func unauthenticate() throws {
    PoilinkSDK.unauthenticate()
  }

  func getRefreshToken(completion: @escaping (Result<String, Error>) -> Void) {
    PoilinkSDK.getRefreshToken(
      onSuccess: { token in completion(.success(token)) },
      onError: { code, msg in completion(.failure(flutterError(code, msg))) }
    )
  }

  func setRefreshToken(appUserId: String, refreshToken: String, completion: @escaping (Result<Void, Error>) -> Void) {
    PoilinkSDK.setRefreshToken(
      appUserId: appUserId,
      refreshToken: refreshToken,
      onSuccess: { completion(.success(())) },
      onError: { code, msg in completion(.failure(flutterError(code, msg))) }
    )
  }

  func progressMission(handle: Int64, missionCode: String, amount: Int64, mode: Int64) throws {
    PoilinkSDK.progressMission(
      missionCode: missionCode,
      amount: Int32(amount),
      mode: ProgressMissionMode(rawValue: Int(mode)) ?? .increase,
      onComplete: nil,
      onError: { [weak self] code, msg in
        self?.eventApi?.onProgressMissionError(handle: handle, errorCode: Int64(code), message: msg) { _ in }
      }
    )
  }

  func progressMissionImmediate(
    missionCode: String,
    amount: Int64,
    mode: Int64,
    idempotencyKey: String?,
    completion: @escaping (Result<[MissionDataMsg], Error>) -> Void
  ) {
    PoilinkSDK.progressMissionImmediate(
      missionCode: missionCode,
      amount: Int32(amount),
      mode: ProgressMissionMode(rawValue: Int(mode)) ?? .increase,
      onSuccess: { list in completion(.success(list.map(toMissionMsg))) },
      onError: { code, msg in completion(.failure(flutterError(code, msg))) },
      idempotencyKey: idempotencyKey
    )
  }

  func getMissionList(filter: MissionListFilterMsg?, completion: @escaping (Result<[MissionDataMsg], Error>) -> Void) {
    let nativeFilter: MissionListFilter?
    if let f = filter {
      nativeFilter = MissionListFilter.buildForNative(
        cycleType: Int32(f.cycleType ?? 0),
        rewardType: Int32(f.rewardType ?? 0),
        progressCode: f.progressCode
      )
    } else {
      nativeFilter = nil
    }
    let list = PoilinkSDK.getMissionList(filter: nativeFilter)
    completion(.success(list.map(toMissionMsg)))
  }

  func showWebPortal(options: WebPortalOptionsMsg) throws {
    PoilinkSDK.showWebPortal(options: buildWebPortalOptions(options))
  }

  func preloadWebPortal(options: WebPortalOptionsMsg) throws {
    PoilinkSDK.preloadWebPortal(options: buildWebPortalOptions(options))
  }

  func closeWebPortal() throws {
    PoilinkSDK.closeWebPortal()
  }

  func syncItemGrants(handle: Int64) throws {
    PoilinkSDK.syncItemGrants(
      onGrants: { [weak self] grants in
        self?.eventApi?.onGrantsPage(handle: handle, grants: grants.map(toGrantMsg)) { _ in }
      },
      onComplete: { [weak self] result in
        self?.eventApi?.onSyncComplete(handle: handle, result: toResultMsg(result)) { _ in }
      },
      onError: { [weak self] code, msg in
        self?.eventApi?.onSyncError(handle: handle, errorCode: Int64(code), message: msg) { _ in }
      }
    )
  }

  private func buildWebPortalOptions(_ msg: WebPortalOptionsMsg) -> PoilinkSDK.WebPortalOptions {
    let handle = msg.handle ?? 0
    let opts = PoilinkSDK.WebPortalOptions()
    if let v = msg.volume { opts.volume = NSNumber(value: v) }
    opts.showMode = PoilinkSDK.ShowMode(rawValue: Int(msg.showMode ?? 0)) ?? .fullscreen
    if let x = msg.embeddedX, let y = msg.embeddedY, let w = msg.embeddedWidth, let h = msg.embeddedHeight {
      opts.embeddedFrame = NSValue(cgRect: CGRect(
        x: CGFloat(x), y: CGFloat(y), width: CGFloat(w), height: CGFloat(h)))
    }
    opts.onShown = { [weak self] in
      self?.eventApi?.onWebPortalShown(handle: handle) { _ in }
    }
    opts.onClose = { [weak self] in
      self?.eventApi?.onWebPortalClose(handle: handle) { _ in }
    }
    opts.onError = { [weak self] code, msg in
      self?.eventApi?.onWebPortalError(handle: handle, errorCode: Int64(code), message: msg) { _ in }
    }
    opts.onMissionChallenge = { [weak self] missionId in
      self?.eventApi?.onMissionChallenge(handle: handle, missionId: missionId) { _ in }
    }
    opts.onRewardReceive = { [weak self] grantId, itemCode, quantity in
      self?.eventApi?.onRewardReceive(
        handle: handle, grantId: grantId, itemCode: itemCode, quantity: Int64(quantity)) { _ in }
    }
    return opts
  }
}

private func flutterError(_ code: Int32, _ message: String) -> PigeonError {
  return PigeonError(code: String(code), message: message, details: nil)
}

private func toMissionMsg(_ m: MissionData) -> MissionDataMsg {
  return MissionDataMsg(
    inAppMissionId: m.inAppMissionId,
    progressCode: m.progressCode,
    title: m.title,
    details: m.details,
    point: m.point,
    targetProgress: Int64(m.targetProgress),
    currentProgress: Int64(m.currentProgress),
    hasAchievement: m.hasAchievement,
    isClaimed: m.isClaimed,
    displayOrder: Int64(m.displayOrder),
    cycleType: Int64(m.cycleType.rawValue),
    rewardType: Int64(m.rewardType.rawValue),
    rewardItemCode: m.rewardItemCode,
    rewardItemQuantity: Int64(m.rewardItemQuantity)
  )
}

private func toGrantMsg(_ m: PendingItemGrant) -> PendingItemGrantMsg {
  return PendingItemGrantMsg(
    grantId: m.grantId,
    itemCode: m.itemCode,
    quantity: Int64(m.quantity),
    grantedAtUnixMs: m.grantedAtUnixMs
  )
}

private func toResultMsg(_ m: SyncItemGrantsResult) -> SyncItemGrantsResultMsg {
  return SyncItemGrantsResultMsg(
    totalSynced: Int64(m.totalSynced),
    totalMarked: Int64(m.totalMarked),
    pageCount: Int64(m.pageCount)
  )
}
